chain/bitcoind_events.go

```diff
@@ -319,3 +354,304 @@ func (b *bitcoindZMQEvents) Stop() error {
+			// determine if there has been a reorg.
+			if info.Blocks <= height {
+				continue
+			}
+
+			// Since we do a height based check, we send
+			// notifications for each block with a height between
+			// the last height we recorded and the new height.
+			for i := height + 1; i <= info.Blocks; i++ {
+				newHash, err := b.client.GetBlockHash(int64(i))
+				if err != nil {
+					log.Errorf("Unable to retrieve "+
+						"block hash: %v", err)
+					continue
+				}
+
+				newBlock, err := b.client.GetBlock(newHash)
+				if err != nil {
+					log.Errorf("Unable to retrieve "+
+						"block: %v", err)
+					continue
+				}
+
+				// notify the client of the new block.
+				b.notifyBlock(newBlock)
+
+				// From our local mempool map, let's remove each
+				// of the transactions that are confirmed in
+				// this new block, since they are no longer in
+				// the mempool.
+				b.mempool.clean(newBlock.Transactions)
+
+				height++
+			}
+
+		case <-b.quit:
+			return
+		}
+	}
+}
+
+// txEventHandlerRPC is a goroutine that uses the RPC client to check the
+// mempool for new transactions.
+func (b *bitcoindRPCPollingEvents) txEventHandlerRPC() {
+	defer b.wg.Done()
+
+	log.Info("Started polling for new bitcoind transactions via RPC.")
+	ticker := time.NewTicker(b.cfg.TxPollingInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			// After each ticker interval, we poll the mempool to
+			// check for transactions we haven't seen yet.
+			txs, err := b.client.GetRawMempool()
+			if err != nil {
+				log.Errorf("Unable to retrieve mempool txs: "+
+					"%v", err)
+				continue
+			}
+
+			// We'll scan through the most recent txs in the
+			// mempool to see whether there are new txs that we
+			// need to send to the client.
+			for _, txHash := range txs {
+				// If the transaction isn't in the local
+				// mempool, we'll send it to the client.
+				if b.mempool.contains(*txHash) {
+					continue
+				}
+
+				// Grab full mempool transaction from hash.
+				tx, err := b.client.GetRawTransaction(txHash)
+				if err != nil {
+					log.Errorf("unable to fetch "+
+						"transaction %s from "+
+						"mempool: %v", txHash, err)
+					continue
+				}
+
+				// Add the transaction to our local mempool.
+				// Note that we only do this after fetching
+				// the full raw transaction from bitcoind.
+				// We do this so that if that call happens to
+				// initially fail, then we will retry it on the
+				// next interval since it is still not in our
+				// local mempool.
+				b.mempool.add(*txHash)
+
+				b.notifyTx(tx.MsgTx())
+			}
+
+		case <-b.quit:
+			return
+		}
+	}
+}
+
+// evictOldTransactions periodically removes any old transactions from our local
+// mempool.
+func (b *bitcoindRPCPollingEvents) evictOldTransactions() {
+	defer b.wg.Done()
+
+	ticker := time.NewTicker(b.cfg.MempoolEvictionAge)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			b.mempool.removeOldTxs(b.cfg.MempoolEvictionAge)
```

Comment:
Why's this needed if we can count on new blocks to actually include those transactions? Also why do we need to add them to a mempool in either case?

Reply:
if we want to api to be the same as the zmq one then the transaction notifications need to include mempool transactions right? if we one sent tx notifications for txs we  get from a block then it is no longer the same as the zmq api. And since we dont know if all txs in the mempool will get confirmed, we need a way to ensure that our mempool doesnt keep expanding. 

Agreed that this is not amazing. Open to suggestions for how to improve it!

Reply:
Hmm, yeah correct that we need the mempool notification to be intact for the `btcwallet` side of things. 

Lemmie take another look at this particular implementation to see if there're areas we can improve.